[GTK][WPE][Skia] Replace SkiaAcceleratedBufferPool by BitmapTexturePool

nikolaszimmermann opened the pull request:
#### 7a790bf39428c4a02fcac4d96a7998855d3b5a35
<pre>
[GTK][WPE][Skia] Replace SkiaAcceleratedBufferPool by BitmapTexturePool
<a href="https://bugs.webkit.org/show_bug.cgi?id=279882">https://bugs.webkit.org/show_bug.cgi?id=279882</a>

Reviewed by Carlos Garcia Campos.

We don&apos;t need SkiaAcceleratedBufferPool, and keep alive
Nicosia::AcceleratedBuffer objects in a pool. Switch to
BitmapTexturePool for GL texture allocation, and let AcceleratedBuffer
receive a BitmapTexture object instead of a SkSurface.

This simplifies the code and lets us re-use the facilities
BitmapTexturePool has to limit resource consumption, etc.

Covered by existing tests.

* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::Buffer::canvas):
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::tryEnsureSurface):
(Nicosia::AcceleratedBuffer::create):
(Nicosia::AcceleratedBuffer::AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::~AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::size const):
(Nicosia::AcceleratedBuffer::tryEnsureSurface):
(Nicosia::AcceleratedBuffer::completePainting):
(Nicosia::AcceleratedBuffer::beginPainting): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
(Nicosia::Buffer::platformWaitUntilPaintingComplete): Deleted.
(Nicosia::Buffer::surface const): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.cpp: Removed.
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.h: Removed.
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintImage):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::CompositingCoordinator):
(WebKit::CompositingCoordinator::invalidate):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.h:

Canonical link: <a href="https://commits.webkit.org/283913@main">https://commits.webkit.org/283913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae52227ec111e4674aaf8869d9a48e99d9663755

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67790 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71845 "Built successfully") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/69908 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18737 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54230 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 wincairo-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70857 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43243 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58615 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/34693 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39917 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/16025 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17289 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/61879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16368 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73543 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11753 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15654 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61681 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11788 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58689 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15050 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9577 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3195 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42979 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/44055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45242 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43794 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -143,35 +150,66 @@ void UnacceleratedBuffer::waitUntilPaintingComplete()
 }
 
-AcceleratedBuffer::AcceleratedBuffer(sk_sp<SkSurface>&& surface, Flags flags)
+AcceleratedBuffer::AcceleratedBuffer(Ref<BitmapTexture>&& texture, Flags flags)
     : Buffer(flags)
+    , m_texture(WTFMove(texture))
```
I think this constructor could just take the BitmapTexture and get the buffer flags from the BitmapTexture flags. It simplifies the caller and ensure there's no mismatch.

nikolaszimmermann replied:
Fixed.

carlosgcampos replied:
I don't see any change here. I meant we don't pass the buffer flags to the constructor, we only pass the texture and pass the buffer flags to the parent by querying the texture flags

nikolaszimmermann replied:
The create() function no longer takes the flags, but computes them, and passes them to AcceleratedBuffer constructor, which forwards them to Buffer constructor.

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -143,35 +150,66 @@ void UnacceleratedBuffer::waitUntilPaintingComplete()
+    auto* canvas = m_surface->getCanvas();
+    canvas->save();
+    canvas->clear(SkColors::kTransparent);
+    return canvas;
```
I know I proposed this, but in this patch it's not obvious why we don't do all this in beginPainting(). It's a bit weird that the save and clear is done here and the restore is done in completePainting. Also the name canvas() sounds like a simple getter. I don't remember why we need the save+clear/restore only for accelerated buffers (I guess because we reuse the buffers from a pool), because my initial idea was to move this to the caller, we return the canvas and you should save+clear/restore or whatever. Also note that completePainting is called even when canvas returns nullptr, but getCanvas() is used there without any null check. If we can move this to the caller (assuming it's harmless for CPU), then we can keep this name as canvas() and only create the surface if it doesn't exist.

nikolaszimmermann replied:
Ok. Will iterate on this.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -68,29 +67,35 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
+            textureFlags.add(BitmapTexture::Flags::SupportsAlpha);
+
+        auto texture = acceleratedBitmapTexturePool->acquireTexture(tileRect.size(), textureFlags);
+        auto buffer = Nicosia::AcceleratedBuffer::create(WTFMove(texture), bufferFlags);
```
```suggestion
        auto buffer = Nicosia::AcceleratedBuffer::create(acceleratedBitmapTexturePool->acquireTexture(tileRect.size(), textureFlags));
```

nikolaszimmermann replied:
Fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -143,41 +150,70 @@ void UnacceleratedBuffer::waitUntilPaintingComplete()
-    m_surface->getCanvas()->save();
-    m_surface->getCanvas()->clear(SkColors::kTransparent);
+    if (auto* canvas = m_surface ? m_surface->getCanvas() : nullptr)
+        return canvas;
```
```suggestion
    if (m_surface)
        return m_surface->getCanvas();
```

carlosgcampos replied:
If the surface doesn't have a canvas we don't set m_surface.

nikolaszimmermann replied:
Bah, sorry, fixing.

nikolaszimmermann replied:
That cannot happen, we always set m_surface. Without m_surface being set, we cannot reach the line 185. We can only get there if someone calls canvas() a second time, which we don't do. But still makes sense to protect against that case.

Still going to fix it to say, `if (m_surface) return m_surface->getCanvas()`, that has the right semantics.

nikolaszimmermann replied:
Ah, ic - there's still the condition that we only ever set m_canvas, if we have "surface->getCanvas()" returning non-zero. I overlooked that. Sorry for the confusion.

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -119,6 +112,20 @@ UnacceleratedBuffer::~UnacceleratedBuffer()
+    // FIXME: ref buffer and unref on release proc?
+    SkSurfaceProps properties = { 0, FontRenderOptions::singleton().subpixelOrder() };
+    m_surface = SkSurfaces::WrapPixels(imageInfo, m_data.get(), imageInfo.minRowBytes64(), &properties);
+    return m_surface ? m_surface->getCanvas() : nullptr;
```
Why do we want to keep the surface alive if it doesn't have a canvas?

carlosgcampos replied:
Ah, I see, to avoid recreating the surface if that's failing to create a canvas.

carlosgcampos commented on Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
```diff
@@ -143,41 +150,66 @@ void UnacceleratedBuffer::waitUntilPaintingComplete()
+Ref<Buffer> AcceleratedBuffer::create(Ref<BitmapTexture>&& texture)
 {
-    return adoptRef(*new AcceleratedBuffer(WTFMove(surface), flags));
+    auto flags = Nicosia::Buffer::Flags { texture->isOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha };
+    return adoptRef(*new AcceleratedBuffer(WTFMove(texture), flags));
```
Missed this, it's ok this way too.

nikolaszimmermann committed:
[GTK][WPE][Skia] Replace SkiaAcceleratedBufferPool by BitmapTexturePool
https://bugs.webkit.org/show_bug.cgi?id=279882

Reviewed by Carlos Garcia Campos.

We don't need SkiaAcceleratedBufferPool, and keep alive
Nicosia::AcceleratedBuffer objects in a pool. Switch to
BitmapTexturePool for GL texture allocation, and let AcceleratedBuffer
receive a BitmapTexture object instead of a SkSurface.

This simplifies the code and lets us re-use the facilities
BitmapTexturePool has to limit resource consumption, etc.

Covered by existing tests.

* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp:
(Nicosia::Buffer::canvas):
(Nicosia::UnacceleratedBuffer::UnacceleratedBuffer):
(Nicosia::UnacceleratedBuffer::tryEnsureSurface):
(Nicosia::AcceleratedBuffer::create):
(Nicosia::AcceleratedBuffer::AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::~AcceleratedBuffer):
(Nicosia::AcceleratedBuffer::size const):
(Nicosia::AcceleratedBuffer::tryEnsureSurface):
(Nicosia::AcceleratedBuffer::completePainting):
(Nicosia::AcceleratedBuffer::beginPainting): Deleted.
* Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.h:
(Nicosia::Buffer::platformWaitUntilPaintingComplete): Deleted.
(Nicosia::Buffer::surface const): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.cpp: Removed.
* Source/WebCore/platform/graphics/skia/SkiaAcceleratedBufferPool.h: Removed.
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
(WebCore::CoordinatedBackingStoreTile::swapBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
(WebCore::CoordinatedGraphicsLayer::paintImage):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::CompositingCoordinator):
(WebKit::CompositingCoordinator::invalidate):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.h:

Canonical link: https://commits.webkit.org/283913@main
